examples/rl-ppo-continuous.rs

```diff
@@ -0,0 +1,200 @@
+    );
+
+    fn forward(&self, x: Tensor2D<BATCH, IN, T>) -> Self::Output {
+        let x = self.l1.forward(x);
+        (
+            self.mu.forward(x),
+            self.std.forward(x),
+            self.value.forward(x),
+        )
+    }
+}
+
+// Hyperparameters stolen from https://github.com/seungeunrho/minimalRL/blob/master/ppo-continuous.py for now
+const LEARNING_RATE: f32 = 0.0003;
+const GAMMA: f32 = 0.9;
+const LAMBDA: f32 = 0.9;
+const EPS_CLIP: f32 = 0.2;
+const K_EPOCH: usize = 10;
+const ROLLOUT_LENGTH: usize = 3;
+const BUFFER_SIZE: usize = 30;
+const MINIBATCH_SIZE: usize = 32;
+
+fn main() {
+    let dev: Cpu = Default::default();
+
+    // initiliaze model - all weights are 0s
+    let mut net: Network<STATE_SIZE, INNER_SIZE, ACTION_SIZE> = nn::BuildModule::build(&dev);
+    let target_net = net.clone();
+
+    let mut optimizer = Adam::new(
+        &net,
+        AdamConfig {
+            lr: LEARNING_RATE,
+            betas: [0.9, 0.999],
+            eps: 1e-08,
+            weight_decay: None,
+        },
+    );
+
+    let mut state: Tensor1D<STATE_SIZE> = init_state();
+
+    // run through training data
+    for _i_epoch in 0..15 {
+        let start = Instant::now();
+
+        // <calc advantage>
+        let (new_state, reward, done) = step_simulation(action);
+
+        let td_target = reward + net.value.forward(net.l1.forward(new_state)) * done * GAMMA;
+        let delta = td_target - net.value.forward(net.l1.forward(state));
+        let delta = delta.array();
+
+        let mut advantage = 0.0;
+        let advantage = delta.take(delta.len() - 1).map(|delta| {
+            advantage = GAMMA * LAMBDA * advantage + delta[0];
+            advantage
+        });
+        // </calc advantage>
+
+        let (mu, std, v) = net.forward(state);
+        //let dist = mu.array().into_iter().zip(std.array()).map(|(mu, std)| Normal::new(mu, std)).collect(); // TODO: is this correct? How should this be done in a more "Tensor"y way?
+
+        let log_prob: Tensor1D<INNER_SIZE> = todo!();
+
+        // ratio = P(action | state, pi_net) / P(action | state, target_pi_net)
+        // but compute in log space and then do .exp() to bring it back out of log space
+        let ratio = (log_prob - old_log_prob).exp();
+
+        // because we need to re-use `ratio` a 2nd time, we need to do some tape manipulation here.
+        let surr1 = ratio.with_empty_tape() * advantage.clone();
+        let surr2 = ratio.clamp(1.0 - EPS_CLIP, 1.0 + EPS_CLIP) * advantage.clone();
+
+        let loss = -surr2.minimum(surr1) + smooth_l1_loss(v, td_target, 1.0);
+        let loss = loss.mean();
+
+        let loss_v = loss.array();
+
+        // TODO: How to nn.utils.clip_grad_norm_(self.parameters(), 1.0)
+
+        // run backprop
+        let gradients = loss.backward();
+
+        // update weights with optimizer
+        optimizer
+            .update(&mut net, gradients)
+            .expect("Unused params");
```

Comment:
"Update" failing does not mean that there were unused parameters, it means that something went wrong allocating memory or executing operations.